presto-main/src/main/java/com/facebook/presto/server/GracefulShutdownHandler.java

```diff
@@ -0,0 +1,126 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.server;
+
+import com.facebook.presto.execution.TaskInfo;
+import com.facebook.presto.execution.TaskManager;
+import io.airlift.bootstrap.LifeCycleManager;
+import io.airlift.log.Logger;
+import io.airlift.units.Duration;
+
+import javax.inject.Inject;
+
+import java.util.List;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static com.google.common.util.concurrent.Uninterruptibles.sleepUninterruptibly;
+import static io.airlift.concurrent.Threads.threadsNamed;
+import static java.lang.Thread.currentThread;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.Executors.newSingleThreadExecutor;
+import static java.util.concurrent.Executors.newSingleThreadScheduledExecutor;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+import static java.util.concurrent.TimeUnit.MINUTES;
+import static java.util.concurrent.TimeUnit.SECONDS;
+
+public class GracefulShutdownHandler
+{
+    private static final Logger log = Logger.get(GracefulShutdownHandler.class);
+    private static final Duration GRACE_PERIOD = new Duration(2, MINUTES);
+    private static final Duration LIFECYCLE_STOP_TIMEOUT = new Duration(30, SECONDS);
+
+    private final ScheduledExecutorService shutdownHandler = newSingleThreadScheduledExecutor(threadsNamed("shutdown-handler-%s"));
+    private final ExecutorService lifeCycleStopper = newSingleThreadExecutor(threadsNamed("lifecycle-stopper-%s"));
+    private final TaskManager sqlTaskManager;
+    private final LifeCycleManager lifeCycleManager;
+    private final AtomicBoolean shutdownRequested = new AtomicBoolean(false);
+
+    @Inject
+    public GracefulShutdownHandler(TaskManager sqlTaskManager, LifeCycleManager lifeCycleManager)
+    {
+        this.sqlTaskManager = requireNonNull(sqlTaskManager, "sqlTaskManager is null");
+        this.lifeCycleManager = requireNonNull(lifeCycleManager, "lifeCycleManager is null");
+    }
+
+    public void requestShutdown()
+    {
+        log.info("Shutdown requested");
+
+        shutdownRequested.set(true);
+
+        //wait for a grace period to start the shutdown sequence
+        shutdownHandler.schedule(() -> {
+            List<TaskInfo> allTasks = sqlTaskManager.getAllTaskInfo();
```

Comment:
I think you need to put all of this `countDownLatch` into a `while (there are any tasks that are not done)` loop, because new tasks might come in after you start this `CountDownLatch`